src/neural/cuda/layers.cc

```diff
@@ -1733,8 +1937,178 @@ AttentionPolicyHead<DataType>::EncoderWeights::~EncoderWeights() {
+                        num_outputs);
+  addBiasBatched(output, output, biases_, 1, batch, num_outputs,
+                 act_, stream);
+}
+
+template <typename DataType>
+AttentionBody<DataType>::AttentionBody(const LegacyWeights& weights,
+                                       void* scratch,
+                                       Activations activations,
+                                       int num_res_blocks, int input_c, int max_batch_size)
+    : embedding_op_size_(weights.ip_emb_b.size()),
+      encoder_head_count_(weights.encoder_head_count),
+      num_resi_blocks_(num_res_blocks),
+      activations_(activations),
+      input_c_(input_c),
+      has_gating_(weights.ip_mult_gate.size() > 0 && weights.ip_add_gate.size() > 0),
+      has_smolgen_(weights.has_smolgen),
+      BaseLayer<DataType>(weights.ip_emb_b.size(), 8, 8, nullptr) {
+
+  allocAndUpload<DataType>(&ip_emb_w_, weights.ip_emb_w, scratch);
+  allocAndUpload<DataType>(&ip_emb_b_, weights.ip_emb_b, scratch);
+
+  {
+    size_t size = 64 * kNumPosEncodingChannels * sizeof(float);
+    ReportCUDAErrors(cudaMalloc(&pos_encoding_, size));
+    ReportCUDAErrors(cudaMemcpy(pos_encoding_, kPosEncoding, size, cudaMemcpyHostToDevice));
+  }
+
+  if (has_gating_) {
+    allocAndUpload<DataType>(&ip_mult_gate_, weights.ip_mult_gate, scratch);
+    allocAndUpload<DataType>(&ip_add_gate_, weights.ip_add_gate, scratch);
+  }
+
+  if (has_smolgen_) {
+    allocAndUpload<DataType>(&smolgen_global_, weights.smolgen_w, scratch);
+    smolgen_global_size_ = 64 * 64;
+  }
+
+  int num_encoders = weights.encoder.size();
+  float alpha = (float) pow(2.0 * num_encoders, 0.25);
+  for (const auto& enc : weights.encoder) {
+    EncoderBlock<DataType>* pW = new EncoderBlock<DataType>(
+        enc, scratch, encoder_head_count_, embedding_op_size_, alpha,
+        smolgen_global_, smolgen_global_size_, max_batch_size,
+        activations_.smolgen_activation, activations_.ffn_activation);
+    encoder_weights_.emplace_back(pW);
+  }
+}
+
+template <typename DataType>
+AttentionBody<DataType>::~AttentionBody() {
+  ReportCUDAErrors(cudaFree(ip_emb_w_));
+  ReportCUDAErrors(cudaFree(ip_emb_b_));
+  ReportCUDAErrors(cudaFree(pos_encoding_));
+  if (has_gating_) {
+    ReportCUDAErrors(cudaFree(ip_mult_gate_));
+    ReportCUDAErrors(cudaFree(ip_add_gate_));
+  }
+  if (has_smolgen_) {
+    ReportCUDAErrors(cudaFree(smolgen_global_));
+  }
+  for (const auto pEnc : encoder_weights_) delete pEnc;
+}
+
+
+template <typename DataType>
+void AttentionBody<DataType>::Eval(
+    int N, DataType* output, const DataType* input, const DataType* input2,
+    void* scratch, size_t scratch_size, cudnnHandle_t /*cudnn*/,
+    cublasHandle_t cublas, cudaStream_t stream) {
+  DataType* scratch0 = (DataType*)scratch;
+  DataType* scratch1 = (DataType*)output;
+  DataType* scratch2 = (DataType*)input2;
+  DataType* scratch3 = scratch2 + scratch_size / (2 * sizeof(DataType));
```

Comment:
Does this have to be relative to input2 or it could have been relative to output as well?